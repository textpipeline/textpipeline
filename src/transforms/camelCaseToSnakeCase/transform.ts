import decamelize from 'decamelize';
import { decamelize as decamelizeProject } from 'oss';

import { Transform } from '../types';

const transform: Transform = {
  name: 'camelCase to snake_case',
  slug: 'camel-to-snake',
  inputTypeName: 'camelCase',
  outputTypeName: 'snake_case',
  openSourceProject: decamelizeProject,
  defaultOutput: '',
  execute: async input => {
    if (!input || !input.trim()) {
      throw new Error('No input detected.');
    }
    return decamelize(input, '_');
  },
};

export default transform;
