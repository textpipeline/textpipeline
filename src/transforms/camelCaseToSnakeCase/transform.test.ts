import camelCaseToSnakeCase from './transform';

describe('camelCaseToSnakeCase', () => {
  it('has the correct attributes', () => {
    expect(camelCaseToSnakeCase).toMatchSnapshot();
  });

  describe('execute', () => {
    it('rejects with a no-input error when supplied an empty string', () =>
      expect(camelCaseToSnakeCase.execute('')).rejects.toEqual(new Error('No input detected.')));

    it('rejects with a no-input error when supplied a string containing only spaces', () =>
      expect(camelCaseToSnakeCase.execute(' ')).rejects.toEqual(new Error('No input detected.')));

    it('rejects with a no-input error when supplied a string containing only tabs', () =>
      expect(camelCaseToSnakeCase.execute('\t')).rejects.toEqual(new Error('No input detected.')));

    it('rejects with a no-input error when supplied a string containing only newlines', () =>
      expect(camelCaseToSnakeCase.execute('\n')).rejects.toEqual(new Error('No input detected.')));

    it('rejects with a no-input error when supplied a string containing only whitespace', () =>
      expect(camelCaseToSnakeCase.execute('\n \t')).rejects.toEqual(new Error('No input detected.')));

    it('returns the original string when given a non-camelCase string', () =>
      expect(camelCaseToSnakeCase.execute('foo')).resolves.toEqual('foo'));

    it('converts a single camelCase word to snake_case', () =>
      expect(camelCaseToSnakeCase.execute('fooBar')).resolves.toEqual('foo_bar'));

    it('converts multiple camelCase words to snake_case', () =>
      expect(camelCaseToSnakeCase.execute('fooBar bazBuz')).resolves.toEqual('foo_bar baz_buz'));

    it('converts camelCase words to snake_case while upholding the original whitespace', () =>
      expect(camelCaseToSnakeCase.execute('fooBar\t \nbazBuz')).resolves.toEqual('foo_bar\t \nbaz_buz'));
  });
});
