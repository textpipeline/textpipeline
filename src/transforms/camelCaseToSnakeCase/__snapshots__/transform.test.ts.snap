// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`camelCaseToSnakeCase has the correct attributes 1`] = `
Object {
  "defaultOutput": "",
  "execute": [Function],
  "inputTypeName": "camelCase",
  "name": "camelCase to snake_case",
  "openSourceProject": Object {
    "description": "Convert a camelized string into a lowercased one with a custom separator: unicornRainbow â†’ unicorn_rainbow",
    "href": "https://github.com/sindresorhus/decamelize",
    "license": Object {
      "href": "https://choosealicense.com/licenses/mit",
      "name": "MIT",
      "slug": "mit",
    },
    "licenseText": "
MIT License

Copyright (c) Sindre Sorhus <sindresorhus@gmail.com> (sindresorhus.com)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the \\"Software\\"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED \\"AS IS\\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
",
    "name": "decamelize",
    "repositoryHref": "https://github.com/sindresorhus/decamelize",
    "slug": "decamelize",
  },
  "outputTypeName": "snake_case",
  "slug": "camel-to-snake",
}
`;
